x-pack/plugins/index_lifecycle_management/public/application/services/policies/frozen_phase.ts

```diff
@@ -0,0 +1,161 @@
+import { isNumber, splitSizeAndUnits } from './policy_serialization';
+import {
+  numberRequiredMessage,
+  PhaseValidationErrors,
+  positiveNumberRequiredMessage,
+} from './policy_validation';
+
+const frozenPhaseInitialization: FrozenPhase = {
+  phaseEnabled: false,
+  selectedMinimumAge: '0',
+  selectedMinimumAgeUnits: 'd',
+  selectedNodeAttrs: '',
+  selectedReplicaCount: '',
+  freezeEnabled: false,
+  phaseIndexPriority: '',
+};
+
+export const frozenPhaseFromES = (phaseSerialized?: SerializedFrozenPhase): FrozenPhase => {
+  const phase = { ...frozenPhaseInitialization };
+  if (phaseSerialized === undefined || phaseSerialized === null) {
+    return phase;
+  }
+
+  phase.phaseEnabled = true;
+
+  if (phaseSerialized.min_age) {
+    const { size: minAge, units: minAgeUnits } = splitSizeAndUnits(phaseSerialized.min_age);
+    phase.selectedMinimumAge = minAge;
+    phase.selectedMinimumAgeUnits = minAgeUnits;
+  }
+
+  if (phaseSerialized.actions) {
+    const actions = phaseSerialized.actions;
+    if (actions.allocate) {
+      const allocate = actions.allocate;
+      if (allocate.require) {
+        Object.entries(allocate.require).forEach((entry) => {
+          phase.selectedNodeAttrs = entry.join(':');
+        });
+        if (allocate.number_of_replicas) {
+          phase.selectedReplicaCount = allocate.number_of_replicas.toString();
+        }
+      }
+    }
+
+    if (actions.freeze) {
+      phase.freezeEnabled = true;
+    }
+
+    if (actions.set_priority) {
+      phase.phaseIndexPriority = actions.set_priority.priority
+        ? actions.set_priority.priority.toString()
+        : '';
+    }
+  }
+
+  return phase;
+};
+
+export const frozenPhaseToES = (
+  phase: FrozenPhase,
+  originalPhase?: SerializedFrozenPhase
+): SerializedFrozenPhase => {
+  if (!originalPhase) {
+    originalPhase = { ...serializedPhaseInitialization };
+  }
+
+  const esPhase = { ...originalPhase };
+
+  if (isNumber(phase.selectedMinimumAge)) {
+    esPhase.min_age = `${phase.selectedMinimumAge}${phase.selectedMinimumAgeUnits}`;
+  }
+
+  esPhase.actions = esPhase.actions ? { ...esPhase.actions } : {};
+
+  if (phase.selectedNodeAttrs) {
+    const [name, value] = phase.selectedNodeAttrs.split(':');
+    esPhase.actions.allocate = esPhase.actions.allocate || ({} as AllocateAction);
+    esPhase.actions.allocate.require = {
+      [name]: value,
+    };
+  } else {
+    if (esPhase.actions.allocate) {
+      // @ts-expect-error
+      delete esPhase.actions.allocate.require;
+    }
+  }
+
+  if (isNumber(phase.selectedReplicaCount)) {
+    esPhase.actions.allocate = esPhase.actions.allocate || ({} as AllocateAction);
+    esPhase.actions.allocate.number_of_replicas = parseInt(phase.selectedReplicaCount, 10);
+  } else {
+    if (esPhase.actions.allocate) {
+      // @ts-expect-error
+      delete esPhase.actions.allocate.number_of_replicas;
+    }
+  }
+
+  if (
```

Comment:
I would love to see these kinds of rules extracted into functions that have descriptive names, a big block comment explaining the intention (with links to relevant docs), and then covered with unit tests to define the desired behavior. I feel like this would make the code easier to scan, and it'd also be easier to dig into them to understand the _why_ behind these rules. Anybody else feel similarly?

Reply:
That would be indeed very helpful 👍

Reply:
I agree, I think we should also move the files to directory called `lib` in that process that lives in `application`. For now, the logic is a carbon-copy of that for cold phase, so at least we know it has been working. I will open an issue for this!